Zstack 3.x.0 support for TI CC2652 chip

leonschenk opened the pull request:
I have been developing this. I have tested with my Sonoff chip, I have no other devices to test.
What needs to be done:
 - bootloader/flashing capability
 - use the device as a router/endpoint capability
 - error handling (unexpected reset)
 - initialization issues (cause not yet investigated)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/zsmartsystems/com.zsmartsystems.zigbee?pullRequest=1352) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/zsmartsystems/com.zsmartsystems.zigbee?pullRequest=1352) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: cdjackson<br/>:white_check_mark: t-8ch<br/>:x: leonschenk<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/zsmartsystems/com.zsmartsystems.zigbee?pullRequest=1352) it.</sub>

Hedda commented:
> What needs to be done:
> 
> * bootloader/flashing capability

@leonschenk  You might already be aware but FYI, many other projects depend on cc2538-bsl (Python cross-platform script) by [JelmerT](https://github.com/JelmerT) to upload firmware via the serial boot loader onto the CC13xx, CC2538 and CC26xx SoC. -> https://github.com/JelmerT/cc2538-bsl (and note it also have old versions but not the newer versions published via PyPI here -> https://pypi.org/project/cc2538/).

And FYI, another related tool is zigpy-cli -> https://github.com/zigpy/zigpy-cli as it can be used to backup the Zigbee network before flashing new firmware and then restoring such Zigbee network backups, and interestingly it saves the backups using the "Open ZigBee Coordinator Backup Format" -> https://github.com/zigpy/open-coordinator-backup which I understand is a unified format which supported under the hood by both Zigbee2MQTT and the ZHA integration in Home Assistant which technically could, in theory, make migration of a Zigbee network possible from one application to another (like from Zigbee2MQTT to OpenHAB ZigBee binding or vice versa).

cdjackson commented:
@Hedda this project also supports firmware updates, and also supports coordinator backups. Ideally these features will be available for all coordinators within this project so that all systems using it have such support.  This isn't really an issue for Home Assistant, or any other system (openHAB etc) - this is a low level library only!

cdjackson commented:
@leonschenk please can you also address the issues raised by `lgtm`. Also, as per our other discussion, please just let me know when you think this is ready for review (ie you're happy that it has reasonable support for errors etc during initialisation).

leonschenk commented:
@cdjackson It will take some time for me to work it out, I will take a look at lgtm. 
@Hedda do you have a device/setup with which you can test the produced library?

Hedda commented:
> @Hedda do you have a device/setup with which you can test the produced library?

Sorry I do not, suggest you perhaps ask for help with testing in openHAB Zigbee Binding request here:

https://github.com/openhab/org.openhab.binding.zigbee/issues/605

and/or maybe also in the openHAB community forum in these two threads:

https://community.openhab.org/t/discussion-about-texas-instruments-z-stack-3-0-and-cc2652-and-cc1352-zigbee-coordinator-adapters-with-openhab-zigbee-binding/129385

https://community.openhab.org/t/iteads-sonoff-zigbee-3-0-usb-dongle-plus-model-zbdongle-p-based-on-texas-instruments-cc2652p-20dbm-radio-mcu-now-sold-for-19-99/126738

leonschenk commented:
@Hedda thank you for your suggestions.
About bootloading capability: I will not be implementing this in the initial setup. If requested, I might implement this later.
About "Open ZigBee Coordinator Backup Format": In this project the backup is independent with the 'dongle' firmware.
I will upload a 'kar' file for testing in the openhab forum. I use version Z-Stack_3.x.0 20220219 from koenkk.

leonschenk commented:
@cdjackson I have some trouble with the discovery of my test device. The device is discovered, but it takes some time before the discovery is complete. This is due to the device which goes to sleep very soon. But is discovery always 'slow' (because I expected a discovery attempt directly after a 'DATA REQUEST', but this is not what I see). Is there any delay, or is it a bug in the Zstack implementation? If you would like to, I am able to send you a wireshark capture.

leonschenk commented:
@cdjackson the dongle works very well.
I have been testing it with 3 different devices for a week now. It has not crashed and works very stable. I have addressed all issues that I had previously.
I hope you will merge my pull request and release a new version.
I will be committed to the work of any bugs or improvements that pop up.

leonschenk commented:
> FYI, Koenkk made a new Zigbee Coordinator firmware release:
> 
> https://github.com/Koenkk/Z-Stack-firmware/tree/master/coordinator/Z-Stack_3.x.0/bin
> 
> https://github.com/Koenkk/Z-Stack-firmware/blob/master/coordinator/Z-Stack_3.x.0/CHANGELOG.md

Thank you @Hedda. I will take a look.

cdjackson commented:
@leonschenk Please can you take a look at the comments I made on  #1280 which I reviewed by mistake. These mostly relate to the changes to the `ZigBeePort` interface and implementations would unless there's a really good reason should not be changed. I think the only other comment I had was around the code generator - just to remove the `@author` tag and add the `@Generated` annotation.

I've not reviewed everything - I'm not sure if there's anything specifically you want me to look at (eg areas where you've made large changes from my original code)? I'm been through all the changed code (eg in common parts of the library and other than the serial interface changes I think it looks ok so would be happy to merge.

Hedda commented:
> @leonschenk Please can you take a look at the comments I made on #1280 which I reviewed by mistake. These mostly relate to the changes to the `ZigBeePort` interface and implementations would unless there's a really good reason should not be changed. I think the only other comment I had was around the code generator - just to remove the `@author` tag and add the `@Generated` annotation.
> 
> I've not reviewed everything - I'm not sure if there's anything specifically you want me to look at (eg areas where you've made large changes from my original code)? I'm been through all the changed code (eg in common parts of the library and other than the serial interface changes I think it looks ok so would be happy to merge.

@leonschenk and @t-8ch Do you have any updates on this PR for Z-Stack 3 driver?

Hedda commented:
@leonschenk and @t-8ch Do you have any updates on/for this PR for Z-Stack 3 driver?

[cdjackson](https://github.com/cdjackson) posted in the forum that this needs to be updated or else it will probably not work:

https://community.openhab.org/t/discussion-about-texas-instruments-z-stack-3-0-and-cc2652-and-cc1352-zigbee-coordinator-adapters-with-openhab-zigbee-binding/129385/11

> "**_Probably the version from November will not work any more - unless you use an old version of OH as well."_**
